modules/indexing/src/test/java/org/apache/ignite/internal/processors/query/stat/StatisticsObsolescenceTest.java

```diff
@@ -18,22 +18,154 @@
+    /** */
+    private long testTimeout;
+
+    /** */
+    @Test
+    public void testObsolescenceWithInsert() throws Exception {
+        doTestObsolescenceUnderLoad(false, 1,
+            key -> sql(String.format("insert into SMALL(A, B, C) values(%d, %d, %d)", key, key, key)));
+    }
+
+    /** */
+    @Test
+    public void testObsolescenceWithUpdate() throws Exception {
+        doTestObsolescenceUnderLoad(true, 0, key -> sql("update SMALL set B=B+1 where A=" + key));
+    }
+
+    /** */
+    @Test
+    public void testObsolescenceWithDelete() throws Exception {
+        doTestObsolescenceUnderLoad(true, -1, key -> sql("delete from SMALL where A=" + key));
+    }
+
+    /** */
+    private void doTestObsolescenceUnderLoad(boolean preload, int rowCntCmp, Consumer<Long> op) throws Exception {
+        // Keep enough data to touch every partition. The statistics collection is sensitive to a partition's empty rows num
+        // and is able to reassemble in this case. This would give false-positive result.
+        int workingRowsNum = RendezvousAffinityFunction.DFLT_PARTITION_COUNT * 10;
+        int preloadCnt = preload ? workingRowsNum : 0;
+
+        int osbInterval = 7;
+
+        testTimeout = osbInterval * 5 * 1000;
+
+        AtomicBoolean stop = new AtomicBoolean();
+
+        try {
+            startGridsMultiThreaded(2);
+
+            for (Ignite ig : G.allGrids())
+                ((IgniteStatisticsManagerImpl)((IgniteEx)ig).context().query().statsManager()).scheduleObsolescence(osbInterval);
+
+            createSmallTable(preloadCnt, null);
+
+            statisticsMgr(0).usageState(StatisticsUsageState.ON);
+            statisticsMgr(0).collectStatistics(buildDefaultConfigurations(SMALL_TARGET));
+
+            // Initialized statistics.
+            assertTrue(waitForCondition(() -> statisticsMgr(0).getLocalStatistics(SMALL_KEY) != null, osbInterval * 1000));
+            assertTrue(waitForCondition(() -> statisticsMgr(1).getLocalStatistics(SMALL_KEY) != null, osbInterval * 1000));
+
+            ObjectStatisticsImpl initStat1 = (ObjectStatisticsImpl)statisticsMgr(0).getLocalStatistics(SMALL_KEY);
+            ObjectStatisticsImpl initStat2 = (ObjectStatisticsImpl)statisticsMgr(1).getLocalStatistics(SMALL_KEY);
+
+            assertEquals(preloadCnt, initStat1.rowCount() + initStat2.rowCount());
+
+            AtomicBoolean enoughDataChanged = new AtomicBoolean(false);
+
+            GridTestUtils.runAsync(() -> {
+                AtomicLong key = new AtomicLong(1L);
+
+                long opCnt = 0;
+
+                while (!stop.get()) {
+                    op.accept(key.getAndIncrement());
+
+                    if (++opCnt == workingRowsNum / 3) {
+                        opCnt = 0;
+
+                        enoughDataChanged.set(true);
+
+                        assertTrue(waitForCondition(() -> stop.get() || !enoughDataChanged.get(), getTestTimeout()));
```

Comment:
I think `CyclicBarrier` is more straitforward way to synchronize two threads here. 
For example we can replace modifiing thread as:
```
                while (!barrier.isBroken()) {
                    op.accept(key.getAndIncrement());

                    if (++opCnt == workingRowsNum / 3) {
                        opCnt = 0;

                        barrier.await();
                        barrier.await();
                    }
                }
```
And remove `enoughDataChanged` and `stop` variables, and several `waitForCondition` calls.
In checking thread we can place 4 `barrier.await()` methods and `barrier.reset()` in the `finally` block.